### ISSUE TITLE
false positives

### ISSUE THREAD

**Issue by du291** [spion/thenlint#2]
There's a couple of common cases when a "no return statement" warning is raised but in fact there is no harm done.
#1 - throwing an exception

```
.then(function () { throw new Error("blah"); });
```
#2 - in tests

```
.then(function () { assert something });
```

**Comment by spion** [comment]
The first one may indeed be an issue.

For the second one you should probably use .done() instead (although I agree that the error isn't quite clear in that case)

**Comment by du291** [comment]
Hi,
thanks for your response. Actually, in tests -- we use mocha -- the test returns the promise like so:

```
it('does something funny', function() {
    return do_something_funny().then(function(result) {
        assert_something
    });
});
```

I believe if the promise was done'd in the test, then the test framework would not recognize assertion failures.
Putting dummy return after the asserts would probably solve the issue but also clutter the tests.

**Comment by spion** [comment]
Oh, I see. Well, this problem along with issue #4 will be impossible to fix without building (or using) a full-fledged static analysis engine, as its impossible to know whether a function call (like an assertion) will throw (or return an actual value) or not without inspecting its entire call graph.

We might look into using TernJS, but ultimately JS is a dynamic language so I'm not sure how fruitful that will be...

For example, its impossible to determine whether the promise consumer will need a value or not without a really good static analysis engine. What if the caller is not mocha but a regular function that expects an actual value? Mocha is even worse in this regard as it would require a special case, because its global - its not imported in any meaningful way that a static analysis engine might understand.

Regarding thrown exceptions that are directly in the function body, I'll have to think about what makes sense and what doesn't. For example, `.then(x => throw new Error("")).then(...)` would probably be a case of a forgotten return statement, while if the second then is replaced with a `catch` it might not be.

Ultimately I think it may come down to choosing between false positives and false negatives. Either that or special-casing testing frameworks (calls to describe/it/etc)

**Comment by du291** [comment]
Hi,
I think I can solve the test issue by not linting my test sources for the moment. For the less-common direct-throw case, perhaps there could be a suppression comment, like in jshint? For example

```
// thenlint returnvalue: false
blahblah.then(throw ...)
```

Also, as for the analysis, I've found out the missing return problem is most severe when there is another promise created inside the function, and without return it is not properly "chained", resulting in race condition. So rather than analyzing if return value is expected, perhaps it's better to analyze whether a new promise is created and not returned. Sometimes, we also tail recurse promises, so yeah, this could turn very ugly. :)
As for the FP/FN choice, if the FP is not a systemic (like with the test) and I have a way to suppress it, then I wouldn't mind getting a few extra FP for less FN.

**Comment by spion** [comment]
How would I detect if a promise was created here?

``` js
var y = f(x);
```

Answer: static analysis engine that inspects the call graph of `f` (And even then, in cases like Bluebird's `promisifyAll` the engine will be unable to analyse the function and determine whether it returns a promise)

As for the suppression comment, if you mean a local one (not a file-level one), `return; //thenlint` would also do nicely.

**Comment by petkaantonov** [comment]
I think then handlers should in all paths either throw or return, otherwise you should use `.tap()` or return some dummy value

**Comment by du291** [comment]
That could work; for the local throw case I'll just put a dummy return (hopefully, it will not upset the other lint) and I'll exclude test sources for now. That (and the closed #3) was the only thing preventing me to use thenlint on a regular basis. To be fair, I benefit from thenlint even as it is now, since it protects against a good number of frequent mistakes. The issue in #4 is rather advanced and I understand it won't be easy to tackle.

**Comment by petkaantonov** [comment]
The return value of the assert can be dummy value too:

``` js
.then(function() {
      return assert(...);
});
```

**Comment by du291** [comment]
Hmm, yea, dummy return after throw works, but upsets a lot of other tools.

```
.then(function () {
                    throw new Error("foo");
                    // thenlint hax
                    //noinspection UnreachableCodeJS,UnnecessaryReturnStatementJS
                    return; // jshint ignore:line
                });
```

**Comment by spion** [comment]
If you're proposing adding suppression comments, PRs are welcome.

You can also give `return void 0;` a try, unless you feel its too much of a hack (it kinda is)
